model/model.go

```diff
@@ -0,0 +1,57 @@
+// Package model is a data modelling interface
+package model
+
+import (
+	"github.com/micro/go-micro/v2/codec"
+	"github.com/micro/go-micro/v2/store"
+)
+
+// Model represents a data model interface for data access
+type Model interface {
+	// Initialise options
+	Init(...Option) error
+	// Retrieve the options
+	Options() Options
+	// Register a data type
+	Register(v interface{}) error
+	// Create an entity
+	Create(v interface{}, opts ...CreateOption) error
+	// Read an entity
+	Read(v interface{}, opts ...ReadOption) error
+	// Update an enity
+	Update(v interface{}, opts ...UpdateOption) error
+	// Delete an entity
+	Delete(v interface{}, opts ...DeleteOption) error
+	// The implementation e.g crud
+	String() string
+}
+
+// Options to the model
+type Options struct {
+	// The store for data storage
+	Store store.Store
+	// The codec for marshalling
+	Codec codec.Marshaler
```

Comment:
I'm not crazy about passing the `micro/go-micro/codec` since that's designed around messages whereas here we're talking about models. 

I'm still arguing in my own head about whether I like the idea of a codec at al or would rather prefer to have a `Marshal/Unmarshal` funcs on the model interface itself. I think I need to have a longer think about possible use cases that would require different marshalling depending on circumstances

Reply:
We tend to pass around interfaces for reuse, its useful to have codecs consolidated in once place. Especially if we're saying types are mostly generated from proto and we want to use the proto codec to serialise. This should ideally be hidden from the user, as in the default codec would be proto and the user would just call model.Create with their type and the model would take care of the encoding for them.

Reply:
We could potentially expose the marshaler on the model but I think since its an internal thing done by the model itself it might not make sense.